android/app/src/main/java/com/microsoft/reactnativecodepush/CodePush.java

```diff
@@ -0,0 +1,412 @@
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.bridge.WritableNativeMap;
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+import com.facebook.react.uimanager.ViewManager;
+import com.facebook.soloader.SoLoader;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+
+public class CodePush {
+
+    private String deploymentKey;
+
+    private boolean resumablePendingUpdateAvailable = false;
+    private boolean didUpdate = false;
+    private Timer timer;
+    private boolean usingTestFolder = false;
+
+    private String assetsBundleFileName;
+
+    private final String FAILED_UPDATES_KEY = "CODE_PUSH_FAILED_UPDATES";
+    private final String PENDING_UPDATE_KEY = "CODE_PUSH_PENDING_UPDATE";
+    private final String PENDING_UPDATE_HASH_KEY = "hash";
+    private final String PENDING_UPDATE_ROLLBACK_TIMEOUT_KEY = "rollbackTimeout";
+    private final String ASSETS_BUNDLE_PREFIX = "assets://";
+    private final String CODE_PUSH_PREFERENCES = "CodePush";
+    private final String DOWNLOAD_PROGRESS_EVENT_NAME = "CodePushDownloadProgress";
+
+    private CodePushPackage codePushPackage;
+    private CodePushReactPackage codePushReactPackage;
+    private CodePushNativeModule codePushNativeModule;
+    private CodePushConfig codePushConfig;
+
+    private Activity mainActivity;
+    private Context applicationContext;
+
+    public CodePush(String deploymentKey, Activity mainActivity) {
+        SoLoader.init(mainActivity, false);
+        this.deploymentKey = deploymentKey;
+        this.codePushPackage = new CodePushPackage(mainActivity.getFilesDir().getAbsolutePath());
+        this.mainActivity = mainActivity;
+        this.applicationContext = mainActivity.getApplicationContext();
+        this.codePushConfig = new CodePushConfig(deploymentKey, this.applicationContext);
+        checkForPendingUpdate(/*needsRestart*/ false);
+    }
+
+    public ReactPackage getReactPackage() {
+        if (codePushReactPackage == null) {
+            codePushReactPackage = new CodePushReactPackage();
+        }
+        return codePushReactPackage;
+    }
+
+    public String getDocumentsDirectory() {
+        return codePushPackage.getDocumentsDirectory();
+    }
+
+    public String getBundleUrl(String assetsBundleFileName) {
+        this.assetsBundleFileName = assetsBundleFileName;
+        String binaryJsBundleUrl = ASSETS_BUNDLE_PREFIX + assetsBundleFileName;
+        try {
+            String packageFile = codePushPackage.getCurrentPackageBundlePath();
+            if (packageFile == null) {
```

Comment:
I think we need to figure out a way to do this before merging this in, since giving precedence to the binary version is important for the app store update scenario.